src/System.Linq/tests/FirstTests.cs

```diff
@@ -0,0 +1,187 @@
+                    where x > Int32.MinValue
+                    select x;
+
+            Assert.Equal(q.First(), q.First());
+        }
+
+        [Fact]
+        public void SameResultsRepeatCallsStringQuery()
+        {
+            var q = from x in new[] { "!@#$%^", "C", "AAA", "", "Calling Twice", "SoS", String.Empty }
+                    where !String.IsNullOrEmpty(x)
+                    select x;
+
+            Assert.Equal(q.First(), q.First());
+        }
+
+        public void TestEmptyIList<T>()
+        {
+            T[] source = { };
+            
+            Assert.NotNull(source as IList<T>);
+            
+            Assert.Throws<InvalidOperationException>(() => source.First());
+        }
+
+        [Fact]
+        public void EmptyIListT()
+        {
+            TestEmptyIList<int>();
+            TestEmptyIList<string>();
+            TestEmptyIList<DateTime>();
+            TestEmptyIList<FirstTests>();
+        }
+
+        [Fact]
+        public void IListTOneElement()
+        {
+            int[] source = { 5 };
+            int expected = 5;
+
+            Assert.NotNull(source as IList<int>);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void IListTManyELementsFirstIsDefault()
+        {
+            int?[] source = { null, -10, 2, 4, 3, 0, 2 };
+            int? expected = null;
+
+            Assert.IsAssignableFrom<IList<int?>>(source);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void IListTManyELementsFirstIsNotDefault()
+        {
+            int?[] source = { 19, null, -10, 2, 4, 3, 0, 2 };
+            int? expected = 19;
+
+            Assert.IsAssignableFrom<IList<int?>>(source);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        private static IEnumerable<T> EmptySource<T>()
+        {
+            yield break;
+        }
+
+        private static void TestEmptyNotIList<T>()
+        {
+            var source = EmptySource<T>();
+
+            Assert.Null(source as IList<T>);
+            
+            Assert.Throws<InvalidOperationException>(() => source.First());
+        }
+
+        [Fact]
+        public void EmptyNotIListT()
+        {
+            TestEmptyNotIList<int>();
+            TestEmptyNotIList<string>();
+            TestEmptyNotIList<DateTime>();
+            TestEmptyNotIList<FirstTests>();
+        }
+
+        [Fact]
+        public void OneElementNotIListT()
+        {
+            IEnumerable<int> source = NumList(-5, 1);
+            int expected = -5;
+
+            Assert.Null(source as IList<int>);
+            
+            Assert.Equal(expected, source.First());
+        }
+
+        [Fact]
+        public void ManyElementsNotIListT()
+        {
+            IEnumerable<int> source = NumList(3, 10);
+            int expected = 3;
+
+            Assert.Null(source as IList<int>);
```

Comment:
Oh. Maybe the answer is to be blatant; change `NumList()` to `NumberRangeGuaranteedNotToBeList()`. That would be pretty self-documenting as to the requirement.

I want to consolidate a lot of what's reused between these once all the tests are done (doing it before would create either order dependencies between the PRs or conflicts between them) so I think I'll do that then.